k9mail/src/main/java/com/fsck/k9/controller/MessageDownloader.java

```diff
@@ -0,0 +1,492 @@
+        fp = new FetchProfile();
+        fp.add(FetchProfile.Item.STRUCTURE);
+        downloadLargeMessages(account, remoteFolder, localFolder, largeMessages, progress, unreadBeforeStart,
+                newMessages, todo, fp);
+        largeMessages.clear();
+
+        Timber.d("SYNC: Synced remote messages for folder %s, %d new messages", folderName, newMessages.get());
+
+        FlagSyncHelper.newInstance(context, controller, syncHelper).refreshLocalMessageFlags(account, remoteFolder,
+                localFolder, syncFlagMessages);
+
+        Timber.d("SYNC: Synced remote messages for folder %s, %d new messages", folderName, newMessages.get());
+
+        if (purgeToVisibleLimit) {
+            localFolder.purgeToVisibleLimit(new MessageRemovalListener() {
+                @Override
+                public void messageRemoved(Message message) {
+                    for (MessagingListener l : controller.getListeners()) {
+                        l.synchronizeMailboxRemovedMessage(account, folderName, message);
+                    }
+                }
+
+            });
+        }
+
+        // If the oldest message seen on this sync is newer than
+        // the oldest message seen on the previous sync, then
+        // we want to move our high-water mark forward
+        // this is all here just for pop which only syncs inbox
+        // this would be a little wrong for IMAP (we'd want a folder-level pref, not an account level pref.)
+        // fortunately, we just don't care.
+        Long oldestMessageTime = localFolder.getOldestMessageDate();
+
+        if (oldestMessageTime != null) {
+            Date oldestExtantMessage = new Date(oldestMessageTime);
+            if (oldestExtantMessage.before(downloadStarted) &&
+                    oldestExtantMessage.after(new Date(account.getLatestOldMessageSeenTime()))) {
+                account.setLatestOldMessageSeenTime(oldestExtantMessage.getTime());
+                account.save(Preferences.getPreferences(context));
+            }
+
+        }
+        return newMessages.get();
+    }
+
+    private <T extends Message> void fetchUnsyncedMessages(final Account account, final Folder<T> remoteFolder,
+            List<T> unsyncedMessages,
+            final List<Message> smallMessages,
+            final List<Message> largeMessages,
+            final AtomicInteger progress,
+            final int todo,
+            FetchProfile fp) throws MessagingException {
+        final String folder = remoteFolder.getName();
+
+        final Date earliestDate = account.getEarliestPollDate();
+        remoteFolder.fetch(unsyncedMessages, fp,
+                new MessageRetrievalListener<T>() {
+                    @Override
+                    public void messageFinished(T message, int number, int ofTotal) {
+                        try {
+                            if (message.isSet(Flag.DELETED) || message.olderThan(earliestDate)) {
+                                if (K9.isDebug()) {
```

Comment:
Maybe extract `processOldOrDeletedMessage()` to reduce the nesting here.

Reply:
Done